File: python/examples/modutil.py
#!/usr/bin/env python3
"""modutil — local cm1756 module utilities (switch / display / LED).
Mirrors c/examples/modutil.c CLI + output format byte-for-byte.

SPDX-License-Identifier: MIT
"""
import argparse
import sys

import bpclient


def main() -> int:
    ap = argparse.ArgumentParser(prog="modutil")
    ap.add_argument("--display", default=None)
    ap.add_argument("--led", type=int, default=-1)
    ap.add_argument("--state", type=int, default=-1)
    ap.add_argument("--led-get", type=int, default=-1, dest="led_get")
    args = ap.parse_args()

    try:
        client = bpclient.Client()
        client.open()
    except Exception:
        sys.stderr.write("client open failed\n")
        return 2
    try:
        client.open_session()
    except Exception:
        pass

    # Switch
    try:
        sw = client.get_switch_position()
        rc = 0
    except Exception as e:
        sw = 0
        rc = bpclient.err_code(e)
    print(f"[switch] rc={rc} ({bpclient.strerror(rc)})  position={sw} (0x{sw:08x})")

    # Display read
    try:
        disp = client.get_display()
        rc = 0
    except Exception as e:
        disp = b"\x00\x00\x00\x00\x00"
        rc = bpclient.err_code(e)
    text = disp[:4].split(b"\x00", 1)[0].decode("ascii", "replace")
    print(f"[display read] rc={rc} ({bpclient.strerror(rc)})  text='{text}' "
          f"(hex {disp[0]:02x} {disp[1]:02x} {disp[2]:02x} {disp[3]:02x})")

    if args.display is not None:
        four = (args.display + "    ")[:4].encode("ascii", "replace")
        try:
            client.set_display(four)
            rc = 0
        except Exception as e:
            rc = bpclient.err_code(e)
        print(f"[display set]  rc={rc} ({bpclient.strerror(rc)})  wrote '{four.decode('ascii')}'")
        try:
            back = client.get_display()
            print(f"[display read-after-write]  text='{back[:4].decode('ascii', 'replace').rstrip(chr(0))}'")
        except Exception:
            pass

    if args.led_get >= 0:
        try:
            state = client.get_led(args.led_get)
            rc = 0
        except Exception as e:
            state = 0
            rc = bpclient.err_code(e)
        print(f"[led get] rc={rc} ({bpclient.strerror(rc)})  id={args.led_get}  "
              f"state={state} (0x{state:08x})")

    if args.led >= 0 and args.state >= 0:
        try:
            client.set_led(args.led, args.state)
            rc = 0
        except Exception as e:
            rc = bpclient.err_code(e)
        print(f"[led set] rc={rc} ({bpclient.strerror(rc)})  id={args.led} "
              f"-> state={args.state}")
        try:
            state = client.get_led(args.led)
            print(f"[led read-after-write] id={args.led}  state={state}")
        except Exception:
            pass

    client.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: c/examples/modutil.c
/*
 * modutil.c — exercise the local cm1756 module utilities:
 *   * Print the current rotary switch position
 *   * Print the current 4-char display contents
 *   * Optionally set the display via --display "ABCD"
 *   * Optionally toggle an LED via --led <id> --state <0|1>
 *   * Print the current LED state of a single LED via --led-get <id>
 *
 * SPDX-License-Identifier: MIT
 */
#include <getopt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bpclient.h"

int main(int argc, char *argv[]) {
    const char *set_display = NULL;
    int led_set_id = -1, led_set_state = -1;
    int led_get_id = -1;

    static struct option opts[] = {
        {"display",  required_argument, 0, 'd'},
        {"led",      required_argument, 0, 'l'},
        {"state",    required_argument, 0, 's'},
        {"led-get",  required_argument, 0, 'g'},
        {0,0,0,0}
    };
    int c, idx;
    while ((c = getopt_long(argc, argv, "d:l:s:g:", opts, &idx)) != -1) {
        if      (c == 'd') set_display   = optarg;
        else if (c == 'l') led_set_id    = (int)strtol(optarg, NULL, 0);
        else if (c == 's') led_set_state = (int)strtol(optarg, NULL, 0);
        else if (c == 'g') led_get_id    = (int)strtol(optarg, NULL, 0);
    }

    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) {
        fprintf(stderr, "client open failed\n"); return 2;
    }
    bp_client_open_session(cl, NULL);

    uint32_t sw = 0;
    int rc = bp_client_get_switch_position(cl, &sw);
    printf("[switch] rc=%d (%s)  position=%u (0x%08x)\n", rc, bp_strerror(rc), sw, sw);

    char disp[5] = {0};
    rc = bp_client_get_display(cl, disp);
    printf("[display read] rc=%d (%s)  text='%s' (hex %02x %02x %02x %02x)\n",
           rc, bp_strerror(rc), disp,
           (uint8_t)disp[0], (uint8_t)disp[1], (uint8_t)disp[2], (uint8_t)disp[3]);

    if (set_display) {
        char four[4] = {' ',' ',' ',' '};
        size_t n = strlen(set_display);
        if (n > 4) n = 4;
        memcpy(four, set_display, n);
        rc = bp_client_set_display(cl, four);
        printf("[display set]  rc=%d (%s)  wrote '%c%c%c%c'\n", rc, bp_strerror(rc),
               four[0], four[1], four[2], four[3]);
        char back[5] = {0};
        if (bp_client_get_display(cl, back) == BP_OK)
            printf("[display read-after-write]  text='%s'\n", back);
    }

    if (led_get_id >= 0) {
        uint32_t s = 0;
        rc = bp_client_get_led(cl, (uint32_t)led_get_id, &s);
        printf("[led get] rc=%d (%s)  id=%d  state=%u (0x%08x)\n",
               rc, bp_strerror(rc), led_get_id, s, s);
    }

    if (led_set_id >= 0 && led_set_state >= 0) {
        rc = bp_client_set_led(cl, (uint32_t)led_set_id, (uint32_t)led_set_state);
        printf("[led set] rc=%d (%s)  id=%d -> state=%d\n", rc, bp_strerror(rc),
               led_set_id, led_set_state);
        uint32_t s = 0;
        if (bp_client_get_led(cl, (uint32_t)led_set_id, &s) == BP_OK)
            printf("[led read-after-write] id=%d  state=%u\n", led_set_id, s);
    }

    bp_client_close(cl);
    return 0;
}
